content/blog/2018/10/2018-10-15-custom-war-packager.adoc

```diff
@@ -0,0 +1,235 @@
+That is why Custom WAR Packager was created in April 2018.
+Later it got support for packaging Docker images,
+and in September 2018 it also got support for Jenkinsfile Runner
+which was created by link:https://github.com/kohsuke/[Kohsuke Kawaguchi]
+and then improved by link:https://github.com/ndeloof[Nicolas de Loof].
+
+== What's inside?
+
+Custom WAR packager is a tool which is available as CLI Executable, Maven Plugin, or Docker package.
+This tool takes input definitions and packages them as requested by the user.
+Everything is managed by a YAML configuration file:
+
+image::/images/post-images/2018-10-15-cwp/cwp_flow.png[Custom WAR Packager build flow]
+
+The tool supports various types of inputs.
+List of plugins can be passed via YAML itself, pom.xml, or a BOM file from jep:309[].
+Custom WAR Packager supports not only there released versions,
+but also builds deployed to the link:/blog/2018/05/15/incremental-deployment/[Incremental repository] (CD flow for Jenkins core and plugins - jep:305[]) and
+even direct builds by Git or directory path specifications.
+It allows building packages from any source, without waiting for official releases.
+The builds are also pretty fast, because the plugin does caching in the local Maven repository by using commit IDs.
+
+Custom WAR packager also supports the following self-configuration options:
+
+** YAML files for link:https://github.com/jenkinsci/configuration-as-code-plugin[Jenkins Configuration as Code]
+** link:https://wiki.jenkins.io/display/JENKINS/Groovy+Hook+Script[Groovy Hooks] (e.g. init hooks for pre-configuration)
+** System properties
+
+== WAR Packaging
+
+WAR packaging happens by default every time the repo is built.
+Generally Custom WAR Packager repackages all inputs into a single WAR file by following conventions defined in the Jenkins core and the JCasC plugin.
+
+Sample configuration:
+
+```yaml
+bundle:
+  groupId: "io.jenkins.tools.war-packager.demo"
+  artifactId: "blogpost-demo"
+  vendor: "Jenkins project"
+  description: "Just a demo for the blogpost"
+war:
+  groupId: "org.jenkins-ci.main"
+  artifactId: "jenkins-war"
+  source:
+    version: 2.138.2
+plugins:
+  - groupId: "io.jenkins"
+    artifactId: "configuration-as-code"
+    source:
+      # Common release
+      version: 1.0-rc2
+  - groupId: "io.jenkins"
+    artifactId: "artifact-manager-s3"
+    source:
+      # Incrementals
+      version: 1.2-rc259.c9d60bf2f88c
+  - groupId: "org.jenkins-ci.plugins.workflow"
+    artifactId: "workflow-job"
+    source:
+      # Git
+      git: https://github.com/jglick/workflow-job-plugin.git
+      commit: 18d78f305a4526af9cdf3a7b68eb9caf97c7cfbc
+  # ...
+systemProperties: {
+    jenkins.model.Jenkins.slaveAgentPort: "9000",
+    jenkins.model.Jenkins.slaveAgentPortEnforce: "true"}
+groovyHooks:
+  - type: "init"
```

Comment:
Why does this have a `type` but `casc` does not?

Reply:
Groovy hooks in Jenkins have types which are a kind of extension point (init, boot-failure). JCasC does not have such equivalent